File: app/src/main/java/com/aus/ausgegeben/util/PeriodUtils.kt
package com.aus.ausgegeben.util

import com.aus.ausgegeben.data.entity.Expense
import com.aus.ausgegeben.ui.isExpense
import com.aus.ausgegeben.ui.isIncome
import com.aus.ausgegeben.ui.isTransfer
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

enum class RecordListPeriod(val label: String) {
    THIS_MONTH("This month"),
    ALL_TIME("All time")
}

enum class AnalyticsPeriod(val label: String, val storageKey: String) {
    THIS_MONTH("This month", "this_month"),
    LAST_MONTH("Last month", "last_month"),
    ALL_TIME("All time", "all_time");

    companion object {
        fun fromStorageKey(key: String?): AnalyticsPeriod =
            entries.find { it.storageKey == key } ?: THIS_MONTH
    }
}

data class AnalyticsPeriodOption(
    val label: String,
    val storageKey: String,
    val rangeMillis: Pair<Long, Long>?
)

/** Inclusive start, exclusive end in local time. Null range = no filtering. */
fun AnalyticsPeriod.dateRangeMillis(nowMillis: Long = System.currentTimeMillis()): Pair<Long, Long>? {
    val cal = Calendar.getInstance().apply { timeInMillis = nowMillis }
    return when (this) {
        AnalyticsPeriod.THIS_MONTH -> monthRange(cal.get(Calendar.YEAR), cal.get(Calendar.MONTH))
        AnalyticsPeriod.LAST_MONTH -> {
            cal.add(Calendar.MONTH, -1)
            monthRange(cal.get(Calendar.YEAR), cal.get(Calendar.MONTH))
        }
        AnalyticsPeriod.ALL_TIME -> null
    }
}

fun analyticsPeriodOptions(
    nowMillis: Long = System.currentTimeMillis(),
    monthsBack: Int = 12
): List<AnalyticsPeriodOption> {
    val cal = Calendar.getInstance().apply {
        timeInMillis = nowMillis
        set(Calendar.DAY_OF_MONTH, 1)
        set(Calendar.HOUR_OF_DAY, 0)
        set(Calendar.MINUTE, 0)
        set(Calendar.SECOND, 0)
        set(Calendar.MILLISECOND, 0)
    }
    return buildList {
        add(
            AnalyticsPeriodOption(
                label = "All time",
                storageKey = AnalyticsPeriod.ALL_TIME.storageKey,
                rangeMillis = null
            )
        )
        repeat(monthsBack.coerceAtLeast(1)) {
            val year = cal.get(Calendar.YEAR)
            val month = cal.get(Calendar.MONTH)
            val range = monthRange(year, month)
            add(
                AnalyticsPeriodOption(
                    label = monthTitle(range.first),
                    storageKey = monthStorageKey(year, month),
                    rangeMillis = range
                )
            )
            cal.add(Calendar.MONTH, -1)
        }
    }
}

fun analyticsPeriodOptionFromStorage(
    key: String?,
    nowMillis: Long = System.currentTimeMillis()
): AnalyticsPeriodOption {
    val normalized = key ?: AnalyticsPeriod.THIS_MONTH.storageKey
    analyticsMonthRangeFromStorageKey(normalized)?.let { range ->
        return AnalyticsPeriodOption(
            label = monthTitle(range.first),
            storageKey = normalized,
            rangeMillis = range
        )
    }
    val legacy = AnalyticsPeriod.fromStorageKey(normalized)
    val range = legacy.dateRangeMillis(nowMillis)
    val storageKey = if (legacy == AnalyticsPeriod.ALL_TIME || range == null) {
        legacy.storageKey
    } else {
        val cal = Calendar.getInstance().apply { timeInMillis = range.first }
        monthStorageKey(cal.get(Calendar.YEAR), cal.get(Calendar.MONTH))
    }
    return AnalyticsPeriodOption(
        label = legacy.displayTitle(nowMillis),
        storageKey = storageKey,
        rangeMillis = range
    )
}

fun analyticsDateRangeMillis(
    storageKey: String,
    nowMillis: Long = System.currentTimeMillis()
): Pair<Long, Long>? {
    if (storageKey == AnalyticsPeriod.ALL_TIME.storageKey) return null
    return analyticsPeriodOptionFromStorage(storageKey, nowMillis).rangeMillis
}

fun AnalyticsPeriod.displayTitle(nowMillis: Long = System.currentTimeMillis()): String = when (this) {
    AnalyticsPeriod.THIS_MONTH -> monthTitle(nowMillis)
    AnalyticsPeriod.LAST_MONTH -> {
        val cal = Calendar.getInstance().apply { timeInMillis = nowMillis }
        cal.add(Calendar.MONTH, -1)
        monthTitle(cal.timeInMillis)
    }
    AnalyticsPeriod.ALL_TIME -> "All time"
}

private fun monthRange(year: Int, month: Int): Pair<Long, Long> {
    val start = Calendar.getInstance().apply {
        set(year, month, 1, 0, 0, 0)
        set(Calendar.MILLISECOND, 0)
    }
    val end = Calendar.getInstance().apply {
        set(year, month, 1, 0, 0, 0)
        set(Calendar.MILLISECOND, 0)
        add(Calendar.MONTH, 1)
    }
    return start.timeInMillis to end.timeInMillis
}

private fun monthStorageKey(year: Int, zeroBasedMonth: Int): String =
    "month:%04d-%02d".format(Locale.US, year, zeroBasedMonth + 1)

private fun analyticsMonthRangeFromStorageKey(key: String): Pair<Long, Long>? {
    if (!key.startsWith("month:")) return null
    val parts = key.removePrefix("month:").split("-")
    val year = parts.getOrNull(0)?.toIntOrNull() ?: return null
    val month = parts.getOrNull(1)?.toIntOrNull()?.minus(1) ?: return null
    if (month !in 0..11) return null
    return monthRange(year, month)
}

private fun monthTitle(millis: Long): String {
    val fmt = SimpleDateFormat("MMMM yyyy", Locale.getDefault())
    return fmt.format(Date(millis))
}

fun List<Expense>.filterByPeriod(period: AnalyticsPeriod, nowMillis: Long = System.currentTimeMillis()): List<Expense> {
    val range = period.dateRangeMillis(nowMillis) ?: return this
    val (start, end) = range
    return filter { it.dateMillis in start until end }
}

data class SpendingInsights(
    val daysLoggedThisWeek: Int = 0,
    val monthExpenseTotal: Double = 0.0,
    val monthIncomeTotal: Double = 0.0,
    val topExpenseCategoryName: String? = null,
    val topExpenseCategoryAmount: Double = 0.0
)

fun recentWeekRangeMillis(nowMillis: Long = System.currentTimeMillis()): Pair<Long, Long> {
    val weekStart = Calendar.getInstance().apply {
        timeInMillis = nowMillis
        add(Calendar.DAY_OF_YEAR, -6)
        set(Calendar.HOUR_OF_DAY, 0)
        set(Calendar.MINUTE, 0)
        set(Calendar.SECOND, 0)
        set(Calendar.MILLISECOND, 0)
    }.timeInMillis
    return weekStart to nowMillis + 1
}

fun computeSpendingInsights(
    monthExpenses: List<Expense>,
    weekExpenses: List<Expense>,
    categoryNames: Map<Long, String>,
): SpendingInsights {
    val daysLogged = weekExpenses
        .map { localDayStartMillis(it.dateMillis) }
        .distinct()
        .size

    val billable = monthExpenses.filter { !it.isTransfer() }
    val monthExpenseTotal = billable.filter { it.isExpense() }.sumOf { it.amount }
    val monthIncomeTotal = billable.filter { it.isIncome() }.sumOf { it.amount }

    val topCategory = billable
        .filter { it.isExpense() }
        .groupBy { it.categoryId }
        .mapValues { (_, items) -> items.sumOf { it.amount } }
        .maxByOrNull { it.value }

    return SpendingInsights(
        daysLoggedThisWeek = daysLogged,
        monthExpenseTotal = monthExpenseTotal,
        monthIncomeTotal = monthIncomeTotal,
        topExpenseCategoryName = topCategory?.key?.let { categoryNames[it] },
        topExpenseCategoryAmount = topCategory?.value ?: 0.0
    )
}

fun computeDayTotals(
    expenses: List<Expense>,
    locale: Locale,
): Map<String, Pair<Double, Double>> {
    val dayTotals = mutableMapOf<Long, Pair<Double, Double>>()
    
    for (expense in expenses) {
        if (expense.isTransfer()) continue
        val dayStart = localDayStartMillis(expense.dateMillis)
        val current = dayTotals.getOrDefault(dayStart, 0.0 to 0.0)
        
        val next = if (expense.isIncome()) {
            (current.first + expense.amount) to current.second
        } else {
            current.first to (current.second + expense.amount)
        }
        dayTotals[dayStart] = next
    }
    
    val dateFormat = SimpleDateFormat("dd.MM EEE", locale)
    return dayTotals.mapKeys { dateFormat.format(Date(it.key)) }
}

File: web/src/utils/periodUtils.ts
import type { AnalyticsPeriodOption, Expense, SpendingInsights } from '@/models/types';
import type { Locale } from '@/i18n';
import { localeTag, t } from '@/i18n';
import { usePreferencesStore } from '@/services/preferencesStore';

function monthRange(year: number, month: number): [number, number] {
  const start = new Date(year, month, 1, 0, 0, 0, 0).getTime();
  const end = new Date(year, month + 1, 1, 0, 0, 0, 0).getTime();
  return [start, end];
}

function resolveLocale(locale?: Locale): string {
  const loc = locale ?? usePreferencesStore.getState().locale;
  return localeTag(loc);
}

function monthTitle(millis: number, locale?: Locale): string {
  return new Intl.DateTimeFormat(resolveLocale(locale), { month: 'long', year: 'numeric' }).format(new Date(millis));
}

function monthStorageKey(year: number, month: number): string {
  return `month:${year}-${String(month + 1).padStart(2, '0')}`;
}

export function localDayStartMillis(now = Date.now()): number {
  const d = new Date(now);
  d.setHours(0, 0, 0, 0);
  return d.getTime();
}

export function thisMonthRange(now = Date.now()): [number, number] {
  const d = new Date(now);
  return monthRange(d.getFullYear(), d.getMonth());
}

export function analyticsPeriodOptions(monthsBack = 24, now = Date.now()): AnalyticsPeriodOption[] {
  const cal = new Date(now);
  cal.setDate(1);
  cal.setHours(0, 0, 0, 0);
  const thisMonth = thisMonthRange(now);
  const options: AnalyticsPeriodOption[] = [
    { label: t('periodAllTime'), storageKey: 'all_time', rangeMillis: null },
    { label: monthTitle(thisMonth[0]), storageKey: 'this_month', rangeMillis: thisMonth },
  ];

  for (let i = 0; i < monthsBack; i++) {
    const year = cal.getFullYear();
    const month = cal.getMonth();
    const range = monthRange(year, month);
    options.push({
      label: monthTitle(range[0]),
      storageKey: monthStorageKey(year, month),
      rangeMillis: range,
    });
    cal.setMonth(cal.getMonth() - 1);
  }

  return options;
}

export function analyticsPeriodOptionFromStorage(storageKey: string, now = Date.now()): AnalyticsPeriodOption {
  const known = analyticsPeriodOptions(24, now).find((option) => option.storageKey === storageKey);
  if (known) return known;

  const range = analyticsDateRangeMillis(storageKey, now);
  if (range) {
    return { label: monthTitle(range[0]), storageKey, rangeMillis: range };
  }

  return { label: t('periodAllTime'), storageKey: 'all_time', rangeMillis: null };
}

export function analyticsDateRangeMillis(storageKey: string, now = Date.now()): [number, number] | null {
  if (storageKey === 'all_time') return null;
  if (storageKey === 'this_month') return thisMonthRange(now);
  if (storageKey.startsWith('month:')) {
    const [, ym] = storageKey.split('month:');
    const [y, m] = ym.split('-').map(Number);
    if (!y || !m) return thisMonthRange(now);
    return monthRange(y, m - 1);
  }
  return thisMonthRange(now);
}

export function formatDateLabel(millis: number, locale?: Locale): string {
  return new Intl.DateTimeFormat(resolveLocale(locale), { weekday: 'short', day: '2-digit', month: 'short' }).format(new Date(millis));
}

export function formatTime(millis: number, locale?: Locale): string {
  return new Intl.DateTimeFormat(resolveLocale(locale), { hour: '2-digit', minute: '2-digit' }).format(new Date(millis));
}

export function dayKey(millis: number): string {
  const d = new Date(millis);
  return `${d.getFullYear()}-${d.getMonth()}-${d.getDate()}`;
}

/** e.g. "Today, 14:32" · "Yesterday, 09:15" · "Mon, 14:32" */
export function formatRelativeTimestamp(millis: number, locale?: Locale, now = Date.now()): string {
  const time = formatTime(millis, locale);
  const day = dayKey(millis);
  const today = dayKey(now);
  if (day === today) return `${t('timeToday')}, ${time}`;

  const yesterday = new Date(now);
  yesterday.setDate(yesterday.getDate() - 1);
  if (day === dayKey(yesterday.getTime())) return `${t('timeYesterday')}, ${time}`;

  const weekAgo = now - 7 * 86_400_000;
  if (millis >= weekAgo) {
    const weekday = new Intl.DateTimeFormat(resolveLocale(locale), { weekday: 'short' }).format(new Date(millis));
    return `${weekday}, ${time}`;
  }

  return `${formatDateLabel(millis, locale)}, ${time}`;
}

export function recentWeekRangeMillis(now = Date.now()): [number, number] {
  const start = new Date(now);
  start.setDate(start.getDate() - 6);
  start.setHours(0, 0, 0, 0);
  return [start.getTime(), now + 1];
}

export function computeSpendingInsights(
  monthExpenses: Expense[],
  weekExpenses: Expense[],
  categoryNames: Map<number, string>,
): SpendingInsights {
  const billableMonth = monthExpenses.filter((e) => e.transactionType !== 'transfer');
  const topCategory = [...billableMonth]
    .filter((e) => e.transactionType === 'expense')
    .reduce((map, expense) => {
      map.set(expense.categoryId, (map.get(expense.categoryId) ?? 0) + expense.amount);
      return map;
    }, new Map<number, number>());

  let topId: number | undefined;
  let topAmount = 0;
  for (const [id, amount] of topCategory) {
    if (amount > topAmount) {
      topId = id;
      topAmount = amount;
    }
  }

  const weekLoggedDays = new Set(
    weekExpenses.map((e) => localDayStartMillis(e.dateMillis)),
  ).size;

  return {
    topCategoryName: topId != null ? categoryNames.get(topId) : undefined,
    topCategoryAmount: topAmount > 0 ? topAmount : undefined,
    weekTotal: weekLoggedDays,
  };
}

export function computeDayTotals(
  expenses: Expense[],
  locale?: Locale,
): Record<string, { income: number; expense: number }> {
  const totals = new Map<string, { income: number; expense: number }>();

  for (const expense of expenses) {
    if (expense.transactionType === 'transfer') continue;
    const label = formatDateLabel(expense.dateMillis, locale);
    const current = totals.get(label) ?? { income: 0, expense: 0 };
    if (expense.transactionType === 'income') {
      totals.set(label, { income: current.income + expense.amount, expense: current.expense });
    } else {
      totals.set(label, { income: current.income, expense: current.expense + expense.amount });
    }
  }

  return Object.fromEntries(totals);
}

/** `YYYY-MM-DD` for `<input type="date">` in local timezone. */
export function toDateInputValue(millis: number): string {
  const d = new Date(millis);
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`;
}

/** Apply a picked local calendar date while preserving time-of-day. */
export function applyLocalDateToMillis(existingMillis: number, dateInput: string): number {
  const [year, month, day] = dateInput.split('-').map((part) => Number.parseInt(part, 10));
  if (!Number.isFinite(year) || !Number.isFinite(month) || !Number.isFinite(day)) return existingMillis;
  const next = new Date(existingMillis);
  next.setFullYear(year, month - 1, day);
  return next.getTime();
}

export function formatCalendarDate(millis: number, locale?: Locale): string {
  return new Intl.DateTimeFormat(resolveLocale(locale), {
    day: '2-digit',
    month: 'short',
    year: 'numeric',
  }).format(new Date(millis));
}
